src/query/service/src/servers/http/v1/session/token_manager.rs

```diff
@@ -0,0 +1,214 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use databend_common_base::base::GlobalInstance;
+use databend_common_cache::Cache;
+use databend_common_cache::LruCache;
+use databend_common_config::InnerConfig;
+use databend_common_exception::ErrorCode;
+use databend_common_exception::Result;
+use databend_common_meta_app::principal::user_token::QueryTokenInfo;
+use databend_common_meta_app::principal::user_token::TokenType;
+use databend_common_meta_app::tenant::Tenant;
+use databend_common_users::UserApiProvider;
+use parking_lot::RwLock;
+use sha2::Digest;
+use sha2::Sha256;
+
+use crate::servers::http::v1::session::token::unix_ts;
+use crate::servers::http::v1::SessionClaim;
+use crate::sessions::Session;
+use crate::sessions::SessionPrivilegeManager;
+
+/// target TTL
+pub const REFRESH_TOKEN_VALIDITY: Duration = Duration::from_hours(4);
+pub const SESSION_TOKEN_VALIDITY: Duration = Duration::from_hours(1);
+
+/// to cove network latency, retry and time skew
+const TOKEN_TTL_DELAY: Duration = Duration::from_secs(300);
+/// in case of client retry, shorted the ttl instead of drop at once
+/// only required for refresh token.
+const TOKEN_DROP_DELAY: Duration = Duration::from_secs(90);
+
+pub struct TokenPair {
+    pub refresh: String,
+    pub session: String,
+}
+
+fn hash_token(token: &[u8]) -> String {
+    hex::encode_upper(Sha256::digest(token))
+}
+pub struct TokenManager {
+    /// store hash only for hit ratio with limited memory, feasible because:
+    /// - token contain all info in itself.
+    /// - for eviction, LRU itself is enough, no need to check expired tokens specifically.
+    session_tokens: RwLock<LruCache<String, ()>>,
+    refresh_tokens: RwLock<LruCache<String, ()>>,
+}
+
+impl TokenManager {
+    pub fn instance() -> Arc<TokenManager> {
+        GlobalInstance::get()
+    }
+
+    #[async_backtrace::framed]
+    pub async fn init(_cfg: &InnerConfig) -> Result<()> {
+        GlobalInstance::set(Arc::new(Self {
+            session_tokens: RwLock::new(LruCache::new(1024)),
+            refresh_tokens: RwLock::new(LruCache::new(1024)),
+        }));
+
+        Ok(())
+    }
+
+    /// used for both issue token for new session and renew token for existing session.
+    /// currently, when renewing, always return a new refresh token instead of update the TTL,
+    /// since now we include expire time in token, which can not be updated.
+    pub async fn new_token_pair(
+        &self,
+        session: &Arc<Session>,
+        old_token_pair: Option<TokenPair>,
```

Comment:
i suppose that it only needs the refresh token to get the new token pair

Reply:
need to delete the old session token?